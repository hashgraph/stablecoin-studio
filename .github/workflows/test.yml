name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test