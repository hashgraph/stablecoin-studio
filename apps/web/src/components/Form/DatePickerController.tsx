import type { StackProps } from '@chakra-ui/react';
import {
	chakra,
	Flex,
	FormControl,
	FormErrorMessage,
	FormLabel,
	HStack,
	IconButton,
	Input,
	InputGroup,
	InputLeftElement,
	Stack,
	Text,
} from '@chakra-ui/react';
import { ChevronLeftIcon, ChevronRightIcon } from '@chakra-ui/icons';
import type { SyntheticEvent } from 'react';
import { forwardRef } from 'react';
import type { Control, UseControllerProps } from 'react-hook-form';
import { Controller } from 'react-hook-form';
import type { ReactDatePickerProps } from 'react-datepicker';
import DatePicker from 'react-datepicker';
import Icon from '../Icon';
import 'react-datepicker/dist/react-datepicker.css';

export interface DatePickerControllerProps extends Omit<ReactDatePickerProps, 'onChange'> {
	control: Control<Record<string, string | number>>;
	dataTestId?: string;
	dateFormat?: string;
	disabled?: boolean;
	isClearable?: boolean;
	isRequired?: boolean;
	label?: string;
	labelProps?: object;
	minimumDate?: Date;
	maximumDate?: Date;
	name: string;
	onChangeAux?: (date: Date | [Date | null, Date | null] | null) => void;
	placeholder?: string;
	rules?: UseControllerProps['rules'];
	showErrors?: boolean;
	containerStyle?: StackProps;
	customHeader?: boolean;
}

const InputStyle = {
	fontSize: '14px',
	_focus: {
		boxShadow: 'none',
	},
	_focusVisible: {
		borderColor: 'inherit',
	},
	_hover: {
		borderColor: 'brand.black',
	},
	_placeholder: {
		color: 'brand.black',
		fontSize: '14px',
	},
	_invalid: {
		borderColor: 'brand.red',
	},
};

const CustomInput = forwardRef<any, any>((props, ref) => {
	return (
		<InputGroup p={2} border='1px solid' borderColor='brand.black' borderRadius='8px'>
			<InputLeftElement h='full' p={3}>
				<Flex justifyContent='center' alignItems='center'>
					<Icon name='CalendarBlank' fontSize={24} color='brand.primary' />
				</Flex>
			</InputLeftElement>
			<Input pl='34px' {...props} ref={ref} variant='unstyled' {...InputStyle} />
		</InputGroup>
	);
});
CustomInput.displayName = 'CustomInput';

const CustomHeader = ({
	date,
	decreaseMonth,
	increaseMonth,
	prevMonthButtonDisabled,
	nextMonthButtonDisabled,
}: any) => {
	return (
		<Stack pb={1} isInline alignItems='center' textAlign='left' pl={4} pr={2}>
			<Text color='gray.700' flex={1} fontSize='sm' fontWeight='medium'>
				{new Intl.DateTimeFormat('en-AU', {
					year: 'numeric',
					month: 'long',
				}).format(date)}
			</Text>
			<IconButton
				borderRadius='full'
				size='sm'
				variant='ghost'
				aria-label='Previous Month'
				icon={<ChevronLeftIcon fontSize='14px' />}
				onClick={decreaseMonth}
				disabled={prevMonthButtonDisabled}
			/>
			<IconButton
				borderRadius='full'
				size='sm'
				variant='ghost'
				aria-label='Next Month'
				icon={<ChevronRightIcon fontSize='14px' />}
				onClick={increaseMonth}
				disabled={nextMonthButtonDisabled}
			/>
		</Stack>
	);
};

const StyledDatePicker = chakra(DatePicker);

const DatePickerController = ({
	control,
	dataTestId,
	dateFormat = 'MMM dd, yyy',
	disabled,
	isClearable = false,
	isRequired = false,
	label,
	labelProps,
	minimumDate,
	maximumDate,
	name,
	onChangeAux,
	placeholder,
	rules,
	showErrors,
	containerStyle,
	customHeader = true,
	...props
}: DatePickerControllerProps) => {
	return (
		<Controller
			control={control}
			name={name}
			rules={rules}
			render={({ field: { onChange, onBlur, value }, fieldState: { invalid, error } }) => {
				const onChangeCustom = (
					date: Date | [Date | null, Date | null] | null,
					event: SyntheticEvent<HTMLInputElement> | undefined,
				) => {
					event?.preventDefault();

					onChange(date);
					onChangeAux && onChangeAux(date);
				};

				return (
					<Stack w='full' {...containerStyle}>
						<FormControl data-testid='form-control' isInvalid={invalid}>
							{label && (
								<FormLabel {...labelProps}>
									<HStack>
										<Text>{label}</Text>
										{isRequired && <Text color='red'>*</Text>}
									</HStack>
								</FormLabel>
							)}
							<StyledDatePicker
								calendarStartDay={1}
								customInput={<CustomInput />}
								data-testid={dataTestId || name}
								dateFormat={dateFormat}
								disabled={disabled}
								dropdownMode='select'
								formatWeekDay={(nameOfDay) => nameOfDay.toString().substring(0, 1)}
								isClearable={isClearable}
								minDate={minimumDate}
								maxDate={maximumDate}
								onChange={(date, e) => onChangeCustom(date, e)}
								onBlur={onBlur}
								onSelect={onBlur}
								placeholderText={placeholder}
								renderCustomHeader={customHeader ? CustomHeader : undefined}
								selected={value ? new Date(value) : undefined}
								showPopperArrow={false}
								{...props}
							/>
							{showErrors && (
								<FormErrorMessage data-testid='datepicker-error-message'>
									{error && error.message}
								</FormErrorMessage>
							)}
						</FormControl>
					</Stack>
				);
			}}
		/>
	);
};

export default DatePickerController;
