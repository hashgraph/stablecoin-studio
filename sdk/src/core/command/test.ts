/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
