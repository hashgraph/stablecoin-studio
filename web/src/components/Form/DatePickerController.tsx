import {
	chakra,
	Flex,
	FormControl,
	FormErrorMessage,
	FormLabel,
	HStack,
	IconButton,
	Input,
	InputGroup,
	InputLeftElement,
	Stack,
	Text,
} from '@chakra-ui/react';
import { ChevronLeftIcon, ChevronRightIcon } from '@chakra-ui/icons';
import { forwardRef, SyntheticEvent } from 'react';
import { Control, Controller, UseControllerProps } from 'react-hook-form';
import DatePicker from 'react-datepicker';
import Icon from '../Icon';
import 'react-datepicker/dist/react-datepicker.css';

export interface DatePickerControllerProps {
	control: Control<Record<string, string | number>>;
	dataTestId?: string;
	dateFormat?: string;
	disabled?: boolean;
	isClearable?: boolean;
	isRequired?: boolean;
	label?: string;
	labelProps?: object;
	minimumDate?: Date;
	name: string;
	onChangeAux?: (date: Date | [Date | null, Date | null] | null) => void;
	placeholder?: string;
	rules?: UseControllerProps['rules'];
	showErrors?: boolean;
}

const CustomInput = forwardRef<any, any>((props, ref) => {
	return (
		<InputGroup p={2} border='1px solid' borderColor='brand.black' borderRadius='8px'>
			<InputLeftElement h='full' p={3}>
				<Flex justifyContent='center' alignItems='center'>
					<Icon name='CalendarBlank' fontSize={24} color='brand.primary' />
				</Flex>
			</InputLeftElement>
			<Input pl='40px' {...props} ref={ref} />
		</InputGroup>
	);
});
CustomInput.displayName = 'CustomInput';

const CustomHeader = ({
	date,
	decreaseMonth,
	increaseMonth,
	prevMonthButtonDisabled,
	nextMonthButtonDisabled,
}: any) => {
	return (
		<Stack pb={1} isInline alignItems='center' textAlign='left' pl={4} pr={2}>
			<Text color='gray.700' flex={1} fontSize='sm' fontWeight='medium'>
				{new Intl.DateTimeFormat('en-AU', {
					year: 'numeric',
					month: 'long',
				}).format(date)}
			</Text>
			<IconButton
				borderRadius='full'
				size='sm'
				variant='ghost'
				aria-label='Previous Month'
				icon={<ChevronLeftIcon fontSize='14px' />}
				onClick={decreaseMonth}
				disabled={prevMonthButtonDisabled}
			/>
			<IconButton
				borderRadius='full'
				size='sm'
				variant='ghost'
				aria-label='Next Month'
				icon={<ChevronRightIcon fontSize='14px' />}
				onClick={increaseMonth}
				disabled={nextMonthButtonDisabled}
			/>
		</Stack>
	);
};

const StyledDatePicker = chakra(DatePicker);

const DatePickerController = ({
	control,
	dataTestId,
	dateFormat = 'MMM dd, yyy',
	disabled,
	isClearable = false,
	isRequired = false,
	label,
	labelProps,
	minimumDate,
	name,
	onChangeAux,
	placeholder,
	rules,
	showErrors,
	...props
}: DatePickerControllerProps) => {
	return (
		<Controller
			control={control}
			name={name}
			rules={rules}
			render={({ field: { onChange, onBlur, value }, fieldState: { invalid, error } }) => {
				const onChangeCustom = (
					date: Date | [Date | null, Date | null] | null,
					event: SyntheticEvent<HTMLInputElement> | undefined,
				) => {
					event?.preventDefault();

					onChange(date);
					onChangeAux && onChangeAux(date);
				};

				return (
					<Stack w='full'>
						<FormControl data-testid='form-control' isInvalid={invalid}>
							{label && (
								<FormLabel {...labelProps}>
									<HStack>
										<Text>{label}</Text>
										{isRequired && <Text color='red'>*</Text>}
									</HStack>
								</FormLabel>
							)}
							<StyledDatePicker
								calendarStartDay={1}
								customInput={<CustomInput />}
								data-testid={dataTestId || name}
								dateFormat={dateFormat}
								disabled={disabled}
								dropdownMode='select'
								formatWeekDay={(nameOfDay) => nameOfDay.substr(0, 1)}
								isClearable={isClearable}
								minDate={minimumDate}
								onChange={(date, e) => onChangeCustom(date, e)}
								onBlur={onBlur}
								onSelect={onBlur}
								placeholderText={placeholder}
								renderCustomHeader={CustomHeader}
								selected={value ? new Date(value) : undefined}
								showPopperArrow={false}
								{...props}
							/>
							{showErrors && (
								<FormErrorMessage data-testid='datepicker-error-message'>
									{error && error.message}
								</FormErrorMessage>
							)}
						</FormControl>
					</Stack>
				);
			}}
		/>
	);
};

export default DatePickerController;
