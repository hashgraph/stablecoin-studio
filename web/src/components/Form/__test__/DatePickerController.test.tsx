import { render } from '../../../test/index';
import type { UseFormReturn } from 'react-hook-form';
import { useForm } from 'react-hook-form';
import { act, fireEvent, waitFor } from '@testing-library/react';
import type { DatePickerControllerProps } from '../DatePickerController';
import DatePickerController from '../DatePickerController';

let form: UseFormReturn;

const RenderWithForm = (props: Omit<DatePickerControllerProps, 'control'>) => {
	const localForm = useForm({
		mode: 'onChange',
		criteriaMode: 'all',
	});

	form = localForm;

	return <DatePickerController control={localForm.control} {...props} />;
};

const defaultProps: Omit<DatePickerControllerProps, 'control'> = {
	name: 'datepicker-test',
	dataTestId: 'Date',
};

const factoryComponent = (props: Partial<DatePickerControllerProps> = defaultProps) => {
	return render(<RenderWithForm name='datepicker-test' {...props} />);
};

describe(`<${DatePickerController.name} />`, () => {
	describe('.render', () => {
		test('should render the component', () => {
			const component = factoryComponent({ label: 'test-label' });

			expect(component.asFragment()).toMatchSnapshot('WithLabel');
		});

		test('should show the label', () => {
			const component = factoryComponent({ label: 'test-label' });

			expect(component.getByText('test-label')).toBeInTheDocument();
		});

		test('should show the error message', async () => {
			const component = factoryComponent({ showErrors: true });

			const errorMessage = 'Expired';

			act(() => {
				form.setError(defaultProps.name, { type: 'expired', message: errorMessage });
			});

			expect(component.getByTestId('datepicker-error-message')).toHaveTextContent(errorMessage);
			expect(component.asFragment()).toMatchSnapshot('WithError');
		});

		test('shouldnt show the error message if showErrors=false', async () => {
			const component = factoryComponent({ showErrors: false });

			const errorMessage = 'Expired';
			act(() => {
				form.setError(defaultProps.name, { type: 'expired', message: errorMessage });
			});

			expect(component.queryByTestId('datepicker-error-message')).toBeNull();
		});

		test('should be disabled if disabled=true', () => {
			const component = factoryComponent({
				disabled: true,
				placeholder: 'datepicker-placeholder',
			});
			const inputDatePicker = component.getByPlaceholderText('datepicker-placeholder');

			expect(inputDatePicker).toBeDisabled();
			expect(component.asFragment()).toMatchSnapshot('Disabled');
		});
	});

	describe('.events', () => {
		test('calls onChangeAux', async () => {
			const onChangeCustom = jest.fn();

			const component = factoryComponent({
				onChangeAux: onChangeCustom,
				placeholder: 'datepicker-placeholder',
			});
			const inputDatePicker = component.getByPlaceholderText('datepicker-placeholder');

			fireEvent.change(inputDatePicker, { target: { value: '12/21/2024' } });

			await waitFor(() => {
				expect(onChangeCustom).toHaveBeenCalled();
			});
		});
	});
});
