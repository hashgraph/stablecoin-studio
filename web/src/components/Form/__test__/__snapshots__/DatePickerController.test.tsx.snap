// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePickerController /> .render should be disabled if disabled=true: Disabled 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="form-control"
      role="group"
    >
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container "
        >
          <span
            aria-live="polite"
            class="react-datepicker__aria-live"
            role="alert"
          />
          <div
            class="chakra-input__group css-1rp2huo"
            data-group="true"
          >
            <div
              class="chakra-input__left-element css-nc17it"
            >
              <div
                class="css-1n2mv2k"
              >
                <svg
                  class="chakra-icon css-g5uxme"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 256 256"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="none"
                    height="256"
                    width="256"
                  />
                  <rect
                    fill="none"
                    height="176"
                    rx="8"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    width="176"
                    x="40"
                    y="40"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="176"
                    x2="176"
                    y1="24"
                    y2="56"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="80"
                    x2="80"
                    y1="24"
                    y2="56"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="40"
                    x2="216"
                    y1="88"
                    y2="88"
                  />
                </svg>
              </div>
            </div>
            <input
              class="chakra-input css-1jo7i1r"
              disabled=""
              id="field-:r4:"
              placeholder="datepicker-placeholder"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<DatePickerController /> .render should render the component: WithLabel 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="form-control"
      role="group"
    >
      <label
        class="chakra-form__label css-1hfoxt1"
        for="field-:r0:"
        id="field-:r0:-label"
      >
        <div
          class="chakra-stack css-84zodg"
        >
          <p
            class="chakra-text css-mgupmj"
          >
            test-label
          </p>
        </div>
      </label>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container "
        >
          <span
            aria-live="polite"
            class="react-datepicker__aria-live"
            role="alert"
          />
          <div
            class="chakra-input__group css-1rp2huo"
            data-group="true"
          >
            <div
              class="chakra-input__left-element css-nc17it"
            >
              <div
                class="css-1n2mv2k"
              >
                <svg
                  class="chakra-icon css-g5uxme"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 256 256"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="none"
                    height="256"
                    width="256"
                  />
                  <rect
                    fill="none"
                    height="176"
                    rx="8"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    width="176"
                    x="40"
                    y="40"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="176"
                    x2="176"
                    y1="24"
                    y2="56"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="80"
                    x2="80"
                    y1="24"
                    y2="56"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="40"
                    x2="216"
                    y1="88"
                    y2="88"
                  />
                </svg>
              </div>
            </div>
            <input
              class="chakra-input css-1jo7i1r"
              id="field-:r0:"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<DatePickerController /> .render should show the error message: WithError 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="form-control"
      role="group"
    >
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container "
        >
          <span
            aria-live="polite"
            class="react-datepicker__aria-live"
            role="alert"
          />
          <div
            class="chakra-input__group css-1rp2huo"
            data-group="true"
          >
            <div
              class="chakra-input__left-element css-nc17it"
            >
              <div
                class="css-1n2mv2k"
              >
                <svg
                  class="chakra-icon css-g5uxme"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 256 256"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    fill="none"
                    height="256"
                    width="256"
                  />
                  <rect
                    fill="none"
                    height="176"
                    rx="8"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    width="176"
                    x="40"
                    y="40"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="176"
                    x2="176"
                    y1="24"
                    y2="56"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="80"
                    x2="80"
                    y1="24"
                    y2="56"
                  />
                  <line
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="16"
                    x1="40"
                    x2="216"
                    y1="88"
                    y2="88"
                  />
                </svg>
              </div>
            </div>
            <input
              aria-describedby="field-:r2:-feedback"
              aria-invalid="true"
              class="chakra-input css-1jo7i1r"
              id="field-:r2:"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        aria-live="polite"
        class="chakra-form__error-message css-10ckq1i"
        data-testid="datepicker-error-message"
        id="field-:r2:-feedback"
      >
        Expired
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
